src/browser/api_protocol/transport_strategy/elipc_strategy.ts

```diff
@@ -205,12 +267,51 @@ export class ElipcStrategy extends ApiTransportBase<MessagePackage> {
 
             if (!e.sender.isDestroyed()) {
                 if (usingBreadcrumbs) {
-                    this.addBreadcrumb(ackObj, messageId, originalPayload.action, 'core/ACK');
+                    ackObj.addBreadcrumb('core/ACK');
                 }
 
                 e.sender.sendToFrame(e.frameRoutingId, electronIpc.channels.CORE_MESSAGE, JSON.stringify(ackObj));
             }
         };
 
     }
+
+    // Handles API batches. Generates a mock of ackDecorator, stores API ACKs, and dispatches as one combined message only after
+    // all API messages in the batch have been resolved (ACK'd)
+    protected ackDeferredDecorator(e: any, messageId: number, originalPayload: any, baseConfiguration: MessageConfiguration): any {
+        const configuration: ElIPCConfiguration = <ElIPCConfiguration>baseConfiguration;
+        const usingBreadcrumbs = configuration.breadcrumbConfiguration.enabled;
+        const deferredAcks: any = [];
+
+        // ACK of the entire API batch
+        const mainAck = this.ackDecorator(e, messageId, originalPayload, baseConfiguration);
+
+        // Stubs the responsibility of a normal ackDecorator
+        // Track state for all ACKs in the batch
+        return (e: any, messageId: number, originalPayload: any): AckFunc => {
+            const ackObj = this.createAckObject(configuration, originalPayload, messageId);
+
+            if (usingBreadcrumbs) {
+                ackObj.addBreadcrumb('core/ackDelegate');
+            }
+
+            // AckFunc to emulate Promise.all()
+            return (payload: any): void => {
+                ackObj.payload = payload;
+
+                if (usingBreadcrumbs) {
+                    ackObj.addBreadcrumb('core/deferredACK');
+                }
+
+                deferredAcks.push(ackObj);
+
+                if (deferredAcks.length === configuration.rendererBatchConfiguration.length) {
+                   mainAck({
+                    success: true,
```

Comment:
What happens if there's a nack? Should this still return success = true? Can the NackPayload still have breadcrumbs added to it?

Reply:
NACK decorators invoke the `ackDecorator`
https://github.com/HadoukenIO/core/blob/00eeebfee7b993c229ae1554c37ffbdfdf4f4e6f/src/browser/api_protocol/transport_strategy/api_transport_base.ts#L47

This is a batch, not a transaction. Success implies the batch was dispatched and consumed by the core.